nixos/modules/virtualisation/digital-ocean-config.nix

```diff
@@ -0,0 +1,165 @@
+          echo "root:$ROOT_PASSWORD" | chpasswd
+          '';
+        unitConfig = {
+          After = [ "network-online.target" ];
+        };
+        serviceConfig = {
+          Type = "oneshot";
+        };
+      };
+
+      /* Set the hostname from Digital Ocean, unless the user configured it in
+       * the NixOS configuration */
+      systemd.services.digitalocean-set-hostname = {
+        enable = hostName == "";
+        path = [ pkgs.curl pkgs.nettools ];
+        description = "Set hostname provided by Digitalocean";
+        wantedBy = [ "network-online.target" ];
+        script = ''
+          set -e
+          DIGITALOCEAN_HOSTNAME=$(curl http://169.254.169.254/metadata/v1/hostname)
+          hostname "$DIGITALOCEAN_HOSTNAME"
+          if [[ ! -e /etc/hostname || -w /etc/hostname ]]; then
+            printf "%s\n" "$DIGITALOCEAN_HOSTNAME" > /etc/hostname
+          fi
+        '';
+        unitConfig = {
+          After =  [ "network-online.target" ];
+        };
+        serviceConfig = {
+          Type = "oneshot";
+        };
+      };
+
+      /* Fetch the ssh keys for root from Digital Ocean */
+      systemd.services.digitalocean-ssh-keys = {
+        enable = cfg.setSshKeys;
+        description = "Set root ssh keys provided by Digital Ocean";
+        wantedBy = [ "network-online.target" ];
+        path = [ pkgs.curl ];
+        script = ''
+          set -e
+          mkdir -m 0700 -p /root/.ssh
+          curl --output /root/.ssh/authorized_keys http://169.254.169.254/metadata/v1/public-keys
+          chmod 600 /root/.ssh/authorized_keys
+        '';
+        serviceConfig = {
+          Type = "oneshot";
+          RemainAfterExit = true;
+        };
+        unitConfig = {
+          ConditionPathExists = "!/root/.ssh/authorized_keys";
+          After = [ "network-online.target" ];
+        };
+      };
+
+      /* Initialize the RNG by running the entropy-seed script from the
+       * Digital Ocean metadata
+       */
+      systemd.services.digitalocean-entropy-seed = {
+        enable = cfg.seedEntropy;
+        description = "Run the kernel RNG entropy seeding script from the Digital Ocean vendor data";
+        wantedBy = [ "network-online.target" ];
+        path = [ pkgs.curl pkgs.mpack ];
+        script = ''
+          set -eo pipefail
+          TEMPDIR=$(mktemp -d)
+          curl http://169.254.169.254/metadata/v1/vendor-data | munpack -tC $TEMPDIR
+          ENTROPY_SEED=$(grep -rl "DigitalOcean Entropy Seed script" $TEMPDIR)
+          ${pkgs.runtimeShell} $ENTROPY_SEED
+          rm -rf $TEMPDIR
+          '';
+        unitConfig = {
+          After = [ "network-online.target" ];
+        };
+        serviceConfig = {
+          Type = "oneshot";
+        };
+      };
+
+    }
+    /* Give nice error messages if a user tries to turn off mutable users and enable
+     * root password or SSH key fetching from Digital Ocean
```

Comment:
Want to remove these defaults+assertions because `mutableUsers=false` does not imply that `/root/.ssh` is readonly, nor even `/etc/shadow`. In fact, enabling these settings is a likely motivation/reason for a server to use `mutableUsers=false` in the first place, so that it can provide auth details in some way other than via the nixos config.